File: spep-lib-java/src/com/qut/middleware/spep/pep/impl/SessionGroupCacheImpl.java
/* 
 * Copyright 2006, Queensland University of Technology
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not 
 * use this file except in compliance with the License. You may obtain a copy of 
 * the License at 
 * 
 *   http://www.apache.org/licenses/LICENSE-2.0 
 * 
 * Unless required by applicable law or agreed to in writing, software 
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT 
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the 
 * License for the specific language governing permissions and limitations under 
 * the License.
 * 
 * Author: Shaun Mangelsdorf
 * Creation Date: 13/12/2006
 * 
 * Purpose: Implements the SessionGroupCache
 */
package com.qut.middleware.spep.pep.impl;

import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;
import java.util.regex.Pattern;

import com.qut.middleware.spep.pep.Messages;
import com.qut.middleware.spep.pep.SessionGroupCache;
import com.qut.middleware.spep.pep.PolicyEnforcementProcessor.decision;
import com.qut.middleware.spep.sessions.PrincipalSession;

/** Implements the SessionGroupCache. */
public class SessionGroupCacheImpl implements SessionGroupCache
{
	private boolean initialized;
	private Map<PrincipalSession, GroupCache> groupCaches;
	private Map<String,List<String>> groupTargets;
	private decision defaultPolicyDecision;
	
	/**
	 * Default constructor
	 * @param defaultPolicyDecision The default policy decision
	 */
	public SessionGroupCacheImpl(decision defaultPolicyDecision)
	{
		if(defaultPolicyDecision == null)
		{
			throw new IllegalArgumentException(Messages.getString("SessionGroupCacheImpl.4")); //$NON-NLS-1$
		}
		
		this.groupCaches = new HashMap<PrincipalSession, GroupCache>();
		this.initialized = false;
		
		if (decision.permit.equals(defaultPolicyDecision) || decision.deny.equals(defaultPolicyDecision))
		{
			this.defaultPolicyDecision = defaultPolicyDecision;
		}
		else
		{
			throw new IllegalArgumentException(Messages.getString("SessionGroupCacheImpl.0")); //$NON-NLS-1$
		}
	}

	/* (non-Javadoc)
	 * @see com.qut.middleware.spep.pep.SessionGroupCache#makeCachedAuthzDecision(com.qut.middleware.spep.sessions.PrincipalSession, java.lang.String)
	 */
	public decision makeCachedAuthzDecision(PrincipalSession principalSession, String resource)
	{
		if (!this.initialized) throw new IllegalArgumentException(Messages.getString("SessionGroupCacheImpl.1")); //$NON-NLS-1$
		
		// Look up group cache for this session.
		GroupCache groupCache;
		
		synchronized(this.groupCaches)
		{
			groupCache = this.groupCaches.get(principalSession);
		}
		
		if (groupCache == null)
		{
			return decision.notcached;
		}
		
		decision result;
		synchronized(groupCache)
		{
			result = groupCache.makeCachedAuthzDecision(resource);
		}
		
		if (result == null)
		{
			return this.defaultPolicyDecision;
		}
		
		return result;
	}
	
	/* (non-Javadoc)
	 * @see com.qut.middleware.spep.pep.SessionGroupCache#clearCache(java.util.Map)
	 */
	public void clearCache(Map<String, List<String>> groupTargetMap)
	{
		synchronized(this.groupCaches)
		{
			this.groupCaches.clear();
			this.groupTargets = groupTargetMap;
			this.initialized = true;
		}
	}
	
	/* (non-Javadoc)
	 * @see com.qut.middleware.spep.pep.SessionGroupCache#clearPrincipalSession(java.lang.String)
	 */
	public void clearPrincipalSession(PrincipalSession principal)
	{
		synchronized(this.groupCaches)
		{
			this.groupCaches.remove(principal);
		}
	}

	/* (non-Javadoc)
	 * @see com.qut.middleware.spep.pep.SessionGroupCache#updateCache(com.qut.middleware.spep.sessions.PrincipalSession, java.lang.String, java.util.List, com.qut.middleware.spep.pep.PolicyEnforcementProcessor.decision)
	 */
	public void updateCache(PrincipalSession principalSession, String groupTarget, List<String> authzTargets,
			decision decision)
	{
		if (!this.initialized) throw new IllegalStateException(Messages.getString("SessionGroupCacheImpl.2")); //$NON-NLS-1$
		
		// Look up group cache for this session.
		GroupCache groupCache;
		synchronized(this.groupCaches)
		{
			groupCache = this.groupCaches.get(principalSession);
			if (groupCache == null)
			{
				groupCache = createDefaultGroupCache();
				this.groupCaches.put(principalSession, groupCache);
			}
		}
		
		synchronized (groupCache)
		{
			groupCache.updateCache(groupTarget, authzTargets, decision);
		}
	}

	
	private class GroupCache
	{
		private Map<String, AuthzTargetCache> authzTargetMap;
		
		protected GroupCache()
		{
			// LinkedHashMap used because it is more efficient at keySet()
			this.authzTargetMap = new LinkedHashMap<String, AuthzTargetCache>();
		}
		
		protected decision makeCachedAuthzDecision(String resource)
		{
			decision result = null;
			
			// Loop through matching group targets
			for (String key : this.authzTargetMap.keySet())
			{
				if (targetMatch(key, resource))
				{
					AuthzTargetCache authzTargetCache = this.authzTargetMap.get(key);
					
					if (authzTargetCache == null)
					{
						result = decision.notcached;
					}
					else
					{
						// Call the AuthzTargetCache to get a cached decision. If null we need to update the cache.
						decision nodeDecision = authzTargetCache.makeCachedAuthzDecision(resource);
						/*if (nodeDecision == null)
						{
							nodeDecision = decision.cache;
						}*/
						
						result = addDecisions(result, nodeDecision);
						
						if (decision.deny.equals(result))
						{
							return result;
						}
					}
				}
			}
			
			return result;
		}
		
		protected void updateCache(String groupTarget, List<String> authzTargets, decision decision)
		{
			// Call the AuthzTargetCache object to update its cache
			AuthzTargetCache authzTargetCache = this.authzTargetMap.get(groupTarget);
			
			if (authzTargetCache == null)
			{
				authzTargetCache = new AuthzTargetCache();
				this.authzTargetMap.put(groupTarget, authzTargetCache);
			}
			
			authzTargetCache.updateCache(authzTargets, decision);
		}
	}
	
	private class AuthzTargetCache
	{
		private Map<String, decision> decisionMap;
		
		protected AuthzTargetCache()
		{
			// LinkedHashMap used because it is more efficient at keySet()
			this.decisionMap = new LinkedHashMap<String, decision>();
		}

		protected decision makeCachedAuthzDecision(String resource)
		{
			decision result = null;
			
			// Loop through all matching targets
			for (String key : this.decisionMap.keySet())
			{
				if (targetMatch(key, resource))
				{
					// Find the cached decision. If null we need to update cache.
					decision nodeDecision = this.decisionMap.get(key);
					if (nodeDecision == null)
					{ 
						nodeDecision = decision.notcached;
					}
					
					// Add the decision to the result
					result = addDecisions(result, nodeDecision);
					
					if (decision.deny.equals(result))
					{
						return result;
					}
				}
			}
			
			return result;
		}
		
		protected void updateCache(List<String> authzTargets, decision decision)
		{
			for (String authzTarget : authzTargets)
			{
				this.decisionMap.put(authzTarget, decision);
			}
		}
	}
	
	protected GroupCache createDefaultGroupCache()
	{
		GroupCache defaultGroupCache = new GroupCache();

		for (Entry<String, List<String>> groupTargetEntry : this.groupTargets.entrySet())
		{
			String groupTarget = groupTargetEntry.getKey();
			List<String> authzTargets = groupTargetEntry.getValue();
			defaultGroupCache.updateCache(groupTarget, authzTargets, decision.notcached);
		}
		
		return defaultGroupCache;
	}
	
	protected boolean targetMatch(String target, String resource)
	{
		if (target.equals(resource)) return true;
		
		Pattern pattern = Pattern.compile(target);
		return pattern.matcher(resource).matches();
	}
	
	protected decision addDecisions(decision lhs, decision rhs)
	{
		// Handle nulls first so we don't get a null pointer.
		if (lhs == null) return rhs;
		if (rhs == null) return lhs;

		switch(lhs)
		{
			case permit:
				/*
				 * Since the following are true we can just return rhs:
				 * permit + permit = permit
				 * permit + deny = deny
				 * permit + cache = cache
				 * permit + error = error
				 */
				return rhs;
				
			case deny:
				/*
				 * deny + permit = deny
				 * deny + deny = deny
				 * deny + cache = deny
				 * deny + error = error
				 */
				if (decision.error.equals(rhs)) return decision.error;
				return decision.deny;
				
			case notcached:
				/*
				 * cache + permit = cache
				 * cache + deny = deny
				 * cache + cache = cache
				 * cache + error = error
				 */
				if (decision.permit.equals(rhs)) return decision.notcached;
				return rhs;
				
			case error:
				/*
				 * error + permit = error
				 * error + deny = error
				 * error + cache = error
				 * error + error = error
				 */
				return decision.error;
		}
		
		throw new IllegalArgumentException(Messages.getString("SessionGroupCacheImpl.3")); //$NON-NLS-1$
	}
}

File: spepcpp/src/spep/pep/impl/SessionGroupCacheImpl.cpp
/* Copyright 2006-2007, Queensland University of Technology
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not 
 * use this file except in compliance with the License. You may obtain a copy of 
 * the License at 
 * 
 *   http://www.apache.org/licenses/LICENSE-2.0 
 * 
 * Unless required by applicable law or agreed to in writing, software 
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT 
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the 
 * License for the specific language governing permissions and limitations under 
 * the License.
 * 
 * Author: Shaun Mangelsdorf
 * Creation Date: 25/09/2006
 * 
 * Purpose: 
 */
 
#include "spep/pep/SessionGroupCache.h"
#include "spep/pep/impl/SessionGroupCacheImpl.h"
#include "spep/Util.h"
#include "spep/UnicodeStringConversion.h"

#include <unicode/regex.h>
#include <unicode/parseerr.h>
#include <unicode/utypes.h>
#include <unicode/ustring.h>

#include <iostream>
#include <sstream>

#include <boost/lexical_cast.hpp>

spep::SessionGroupCacheImpl::SessionGroupCacheImpl(saml2::Logger *logger, spep::Decision defaultPolicyDecision) :
    mLocalLogger(logger, "spep::SessionGroupCacheImpl"),
    mCacheMutex(),
    mInitialized(false),
    mDefaultPolicyDecision(defaultPolicyDecision)
{
}

spep::SessionGroupCacheImpl::~SessionGroupCacheImpl()
{
    for (auto iter : mGroupCaches) {
        delete iter.second;
    }
}

bool spep::SessionGroupCacheImpl::targetMatch(const UnicodeString &pattern, const UnicodeString &target)
{
    // TODO Opportunity for caching of compiled regex patterns is here.
    UParseError parseError;
    UErrorCode errorCode = U_ZERO_ERROR;
    // Perform the regular expression matching here.
    UBool result = RegexPattern::matches(pattern, target, parseError, errorCode);

    if (U_FAILURE(errorCode))
    {
        // TODO throw u_errorName( errorCode )
        throw std::exception();
    }

    // FALSE is defined by ICU. This line for portability.
    return (result != FALSE);
}

spep::GroupCache *spep::SessionGroupCacheImpl::createDefaultGroupCache()
{
    // Create a new group cache
    GroupCache *groupCache = new GroupCache();

    ScopedLock lock(mCacheMutex);

    // Populate the cache with the group targets
    for (auto iter: mGroupTargets) {
        groupCache->updateCache(iter.first, iter.second, Decision::CACHE);
    }

    return groupCache;
}

void spep::SessionGroupCacheImpl::updateCache(const std::wstring &sessionID, const UnicodeString& groupTarget, std::vector<UnicodeString> &authzTargets, spep::Decision decision)
{
    ScopedLock lock(mCacheMutex);

    // operator[] here is ok because we're creating the entry anyway.
    GroupCache *groupCache = mGroupCaches[sessionID];
    if (groupCache == NULL)
    {
        // If the group cache doesn't exist, create it and insert into the map.
        groupCache = createDefaultGroupCache();
        mGroupCaches[sessionID] = groupCache;
    }

    // Perform a cache update.
    groupCache->updateCache(groupTarget, authzTargets, decision);
}

void spep::SessionGroupCacheImpl::clearCache(std::map< UnicodeString, std::vector<UnicodeString> > &groupTargets)
{
    ScopedLock lock(mCacheMutex);

    for (auto iter = mGroupCaches.begin(); iter != mGroupCaches.end(); ++iter)
    {
        // deleting these while they're still in the map would be bad.
        // but we are inside a lock so it doesn't matter.
        GroupCache *groupCache = iter->second;
        delete groupCache;
    }

    mGroupCaches.clear();

    mGroupTargets.clear();
    mGroupTargets = groupTargets;

    mLocalLogger.debug() << "Cleared cache. " << boost::lexical_cast<std::string>(groupTargets.size()) << " group targets in cache now.";
    for (auto groupTargetIterator = groupTargets.begin();
        groupTargetIterator != groupTargets.end(); ++groupTargetIterator)
    {
        mLocalLogger.debug() << "Group target " << UnicodeStringConversion::toString(groupTargetIterator->first);
        for (auto authzTargetIterator = groupTargetIterator->second.begin();
            authzTargetIterator != groupTargetIterator->second.end(); ++authzTargetIterator)
        {
            mLocalLogger.debug() << "- authz target " << UnicodeStringConversion::toString(*authzTargetIterator);
        }
    }

    // TODO This was previously in updateCache() as the last line. Why?
    mInitialized = true;
}

spep::Decision spep::SessionGroupCacheImpl::makeCachedAuthzDecision(const std::wstring& sessionID, const UnicodeString& resource)
{
    {
        std::stringstream ss;
        ss << "Going to make cached authz decision for session [" << UnicodeStringConversion::toString(sessionID);
        ss << "] resource: " << UnicodeStringConversion::toString(resource) << std::ends;

        mLocalLogger.debug() << ss.str();
    }

    if (!mInitialized)
    {
        mLocalLogger.error() << "Session group cache has not been initialized. Rejecting authorization request.";
        return Decision::ERROR;
    }

    ScopedLock lock(mCacheMutex);

    // Find the group cache for this session.
    auto iter = mGroupCaches.find(sessionID);
    if (iter == mGroupCaches.end() || iter->second == NULL)
    {
        // Not there, we need a new set of authz decisions from the PDP.
        mLocalLogger.debug() << "No authorization cache for this session. Returning.";
        return Decision::CACHE;
    }

    GroupCache *groupCache = iter->second;
    mLocalLogger.debug() << "Got authorization cache for session. Going to perform cached authorization.";

    // Go into the group cache to make the decision.
    Decision result = groupCache->makeCachedAuthzDecision(resource);

    // No decision was made, so return the default policy decision.
    if (result == Decision::NONE)
    {
        std::stringstream ss;
        ss << "No matching policy for resource: " << UnicodeStringConversion::toString(resource) << ". Falling back to default policy decision." << std::ends;
        mLocalLogger.debug() << ss.str();
        return mDefaultPolicyDecision;
    }

    {
        std::stringstream ss;
        ss << "Resource [" << UnicodeStringConversion::toString(resource) << "] <- session [" << UnicodeStringConversion::toString(sessionID) << "] .. result is ";

        if (result == Decision::PERMIT)
        {
            ss << "PERMIT";
        }
        else if (result == Decision::DENY)
        {
            ss << "DENY";
        }
        else if (result == Decision::CACHE)
        {
            ss << "CACHE";
        }
        else if (result == Decision::ERROR)
        {
            ss << "ERROR";
        }
        else
        {
            ss << "An invalid decision";
        }

        ss << std::ends;
        mLocalLogger.info() << ss.str();
    }

    return result;
}

spep::AuthzTargetCache::AuthzTargetCache()
{
}

spep::AuthzTargetCache::~AuthzTargetCache()
{
}

spep::Decision spep::AuthzTargetCache::makeCachedAuthzDecision(const UnicodeString& resource)
{
    Decision result;

    // Loop through the cache of decisions
    for (auto iter = mDecisions.begin(); iter != mDecisions.end(); ++iter)
    {
        // If the resource matches this authz target..
        if (SessionGroupCacheImpl::targetMatch(iter->first, resource))
        {
            // Get the decision
            Decision nodeDecision = iter->second;
            if (nodeDecision == Decision::NONE)
            {
                // If it's empty, the cache needs to be updated.
                nodeDecision = Decision::CACHE;
            }

            // Add the decision at this node to the resulting decision
            result += nodeDecision;

            // If it's a deny, we can fail-fast
            if (result == Decision::DENY)
            {
                return result;
            }
        }
    }

    return result;
}

void spep::AuthzTargetCache::updateCache(std::vector<UnicodeString> &authzTargets, spep::Decision decision)
{
    // Set all the authz target decisions to this decision
    std::vector<UnicodeString>::iterator iter;
    for (iter = authzTargets.begin(); iter != authzTargets.end(); ++iter)
    {
        mDecisions[*iter] = decision;
    }
}

spep::GroupCache::GroupCache()
{
}

spep::GroupCache::~GroupCache()
{
    // Delete all the authz target caches from the map.
    AuthzTargetMap::iterator iter;
    for (iter = mAuthzTargets.begin(); iter != mAuthzTargets.end(); ++iter)
    {
        delete iter->second;
    }
}

spep::Decision spep::GroupCache::makeCachedAuthzDecision(const UnicodeString& resource)
{
    Decision result;

    // Loop through authz target caches
    for (auto iter = mAuthzTargets.begin(); iter != mAuthzTargets.end(); ++iter)
    {
        // If the group target matches..
        if (SessionGroupCacheImpl::targetMatch(iter->first, resource))
        {
            AuthzTargetCache *authzTargetCache = iter->second;

            if (nullptr == authzTargetCache)
            {
                // No authz target cache there, we need to update the cache.
                result = Decision::CACHE;
            }
            else
            {
                // Get the decision from the authz target cache
                Decision nodeDecision = authzTargetCache->makeCachedAuthzDecision(resource);

                // Add it to the result decision.
                result += nodeDecision;

                // If it's a deny, we can fail fast.
                if (result == Decision::DENY)
                {
                    return result;
                }
            }
        }
    }

    return result;
}

void spep::GroupCache::updateCache(const UnicodeString& groupTarget, std::vector<UnicodeString> &authzTargets, spep::Decision decision)
{
    AuthzTargetCache *authzTargetCache = mAuthzTargets[groupTarget];

    if (authzTargetCache == nullptr)
    {
        authzTargetCache = new AuthzTargetCache();
        mAuthzTargets[groupTarget] = authzTargetCache;
    }

    authzTargetCache->updateCache(authzTargets, decision);
}
